llvm/test/CodeGen/AMDGPU/back-off-barrier-subtarget-feature.ll

```diff
@@ -45,6 +45,7 @@ define void @back_off_barrier_no_fence(ptr %in, ptr %out) #0 {
 ; GFX11-BACKOFF-NEXT:    s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
 ; GFX11-BACKOFF-NEXT:    flat_load_b32 v0, v[0:1]
 ; GFX11-BACKOFF-NEXT:    s_waitcnt vmcnt(0) lgkmcnt(0)
+; GFX11-BACKOFF-NEXT:    s_waitcnt_vscnt null, 0x0
```

Comment:
I am also confused by `setNonKernelFunctionInitialState`. It does not seem like it should be necessary for this patch.

Reply:
Without it there's quite a few more changes, e.g.
- MSG_DEALLOC_VGPRS is back in `llvm.amdgcn.set.inactive.chain.arg.ll`
- `atomicrmw-expand.ll` has a few missing `s_waitcnt_vscnt null, 0x0` which I'm not sure if it's good or bad

Given that the PSDB passes with the changes I assume they're here for a reason, but Juan cannot tell us why right now so I don't know what the reasoning is

Reply:
I think the effect of `setNonKernelFunctionInitialState` is, roughly, to say that all the wait counters are in an unknown state, instead of known to be 0. It seems very odd to do this immediately after emitting `s_waitcnt 0`.

Reply:
Is this fine for now then?

Reply:
In this patch I would prefer that `setNonKernelFunctionInitialState` only sets vscnt to unknown, and leaves the other counters as known to be zero. Does that change affect the tests?

Also I think it might need to do `Brackets.setPendingEvent(SCRATCH_WRITE_ACCESS)` (like in #73122) to indicate that there _might_ be outstanding scratch writes. Does that affect the tests? I'm not sure if it also needs to set the other pending event bits too.

Reply:
> In this patch I would prefer that setNonKernelFunctionInitialState only sets vscnt to unknown, and leaves the other counters as known to be zero. Does that change affect the tests?

With
```
    setScoreUB(VS_CNT, getWaitCountMax(VS_CNT));
    PendingEvents |= WaitEventMaskForInst[VS_CNT];
````

Only `CodeGen/AMDGPU/vgpr-descriptor-waterfall-loop-idom-update.ll` changes, everything else stays the same

I have no strong preference on whether to finish this first or the other patch - I just picked it up and I'm still learning about InsertWaitCnt myself. Though, I think if this patch has a lot more improvements than regressions we should land it first and add a TODO for the remaining bad cases

Reply:
Hello!

In an initial version of the patch, before we introduced de soft-waitcnts, setting the counters to 0 in the initial state would end up removing the "s_waitcnt 0" placed at non-kernel function entry. This is not needed anymore since we make the distinction between soft/strong waitcnts. However, we still have to do it for vscnt as jay said.